### ISSUE TITLE
Problems launching widget from another widget

### ISSUE THREAD

**Issue by theJuddson** [ozone-development/ozp-iwc#394]
This is an extension of bug #385. I have read as much documentation as I can find surrounding widget/application/listing launching, but have failed to get it working in my local instance with my custom listings.

I added a custom listing to my OZP instance, but there didn't seem to be a way to assign a "unique name" to the listing. I noticed that there's a "unique_name" column on the listing table, so I updated that column with a name that seemed reasonable. 

Here's the database entry for my map app:
```
id |           title            |         approved_date         |          edited_date          |        description         |                    launch_url                    | version_name | unique_name | what_is_new | description_short | requirements | approval_stat
us | is_enabled | is_featured | avg_rate | total_votes | total_rate5 | total_rate4 | total_rate3 | total_rate2 | total_rate1 | total_reviews | iframe_compatible | security_marking | is_private | agency_id | banner_icon_id | current_rejection_id | large_banner_icon_id |
large_icon_id | last_activity_id | listing_type_id | required_listings_id | small_icon_id
----+----------------------------+-------------------------------+-------------------------------+----------------------------+--------------------------------------------------+--------------+-------------+-------------+-------------------+--------------+--------------
---+------------+-------------+----------+-------------+-------------+-------------+-------------+-------------+-------------+---------------+-------------------+------------------+------------+-----------+----------------+----------------------+----------------------+-
--------------+------------------+-----------------+----------------------+---------------
1 | Map App                    | 2016-02-25 15:20:21.765774+00 | 2016-02-25 15:20:21.792406+00 | Map Application        | http://localhost:8080                            | 1.0          | map.app     |             | Map               | None         | APPROVED
   | t          | f           |        0 |           0 |           0 |           0 |           0 |           0 |           0 |             0 | t                 | UNCLASSIFIED     | f          |         1 |              4 |                      |                    5 |
            3 |                8 |               1 |                      |             2
```

From a separate custom listing/application, also connected to the IWC Bus, I execute the following JavaScript on a button click:
```javascript
var locationRequest = new iwc.data.Reference("map.location.request");
var mapOpenRef = new iwc.data.Reference("/applications/open/map.app");
mapOpenRef.get().then(function(val){
    console.log("Map already open. Sending Location Request.");
    locationRequest.set(message);
}).catch(function(err){
    console.log("Map closed. Opening. Error: %O", err);
    var mapRef = new iwc.system.Reference("/application/map.app");
    mapRef.launch()
        .then(function () {
            console.log("Map opened successfully. Sending Location Request.");
            locationRequest.set(message);
        })
        .catch(function(e){
            locationRequest.set(message);
            console.log("Failed to open map. Sending Location Request anyways. Error: %O", e);
        });
});
```

Which results in the following console output:
```
Map closed. Opening. Error: noResource
Failed to open map. Sending Location Request anyways. Error: noResource
```
Note that this output is the same regardless of the open state of the map listing, however if the map listing is open, the locationRequest.set() functionality does succeed. My problem is that I cannot seem to be able to get a reference to my "map.app" listing to view the status of or launch that listing.

Is this a bug in IWC, or user error on my part?

**Comment by Kevin-K** [comment]
Interesting,

The IWC distributables includes a web-app for debugging/checking api resource values. Iit should be reachable at the `/debugger` endpoint of where you are hosting the IWC (ex:http://localhost:13000/debugger)

If you navigate to the `System API` tab you should get a table view of all the Applications the IWC has parsed from your DB. Is the `/application/map.app` visible there?


Alternatively, if you want to get a list of all applications while developing, you can run the following in your application:
```
var apps = new iwc.system.Reference('/application');
apps.list().then(function(apps){ 
   console.log(apps);
});
```


Before version `1.2.2` application referencing was done based on its database ID, but was moved to use the `unique_name`, which from your db output seems to match `map.app` so `/application/map.app` should be a valid resource.

Let me know if the application is visible under `/application` from one of the 2 methods above. The github hosted IWC unfortunately doesn't interact with a DB otherwise I would point to examples from there.

**Comment by theJuddson** [comment]
I explored the system.api as well, which is how I discovered ticket #385. I haven't been able to pull any data back from that API. As a sanity check it just re-cloned the ozp-anzible repo and deployed a fresh VM, following developer "quick start" setup instructions. 

1. `git clone https://github.com/ozone-development/ozp-ansible.git`
2. `cd ozp-ansible`
3. `mv group_vars/all/vault_unencrypted.yml group_vars/all/vault.yml`
4. `vagrant up` 

After the VM finished installing / initializing, I ssh'ed into the VM (vagrant ssh) and updated the Bread Basket index.html (/usr/local/ozp/frontend/demo_apps/centerSampleListings/breadBasket/index.html) to add some JavaScript:
```html
<!DOCTYPE html>
<html>
  <head>
    <link href="../../bower_components/bootstrap/dist/css/bootstrap.min.css" rel="stylesheet">
    <script src="../../bower_components/jquery/dist/jquery.min.js"></script>
    <script src="../../bower_components/bootstrap/dist/js/bootstrap.min.js"></script>
    <!-- Start custom content (theJuddson) -->
    <script src="../../bower_components/ozp-iwc/dist/js/ozpIwc-client.js"></script>
    <script>
        $(document).ready(function(){
            var peer = getParameterByName("ozpIwc.peer");
            var base = peer.replace("/iwc", "");
            console.log("Peer URL: " + peer);
            var iwc = new ozpIwc.Client({peerUrl: peer});
            iwc.connect().then(
                function () { //Success
                    console.log("Connected to IWC!")
                    var apps = new iwc.system.Reference('/application');
                    apps.list().then(function(apps){
                            console.log("Applications retrieved via IWC: %O", apps);
                    });
                    var am = new iwc.system.Reference('/application/ozp.test.air_mail');
                    am.launch().then(function(){
                            console.log("Success!");
                    });
                });
            $.ajax(base + "api/self/library/", { contentType: "application/json" }).success(function(data){console.log("Applications retrieved via REST Api: %O", data);});
        });

        function getParameterByName(name, url) {
            if (!url) url = window.location.href;
            name = name.replace(/[\[\]]/g, "\\$&");
            var regex = new RegExp("[?&]" + name + "(=([^&#]*)|&|#|$)"),
            results = regex.exec(url);
            if (!results) return null;
            if (!results[2]) return '';
            return decodeURIComponent(results[2].replace(/\+/g, " "));
        }
    </script>
    <!-- End custom content (theJuddson) -->
  </head>

  <body>
    <div class="container-fluid" style="background-color: #8D1919; color: #CCCCCC;">
      <div class="row">
        <div class="col-md-12">
          <h1> Bread Basket </h1>
          <p> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed dictum
            lectus faucibus ante dapibus commodo.
          </p>
        </div>
      </div>
      <div class="row">
        <div class="col-md-10">
          <img src="baguettes.svg" width="50%" height="50%">
        </div>
      </div>
    </div>
  </body>
</html>
```

I then logged in to https://localhost:4440/hud with wsmith/password (using basic auth), and bookmarked the Bread Basket and Air Mail listings.

Adding Bread Basket to my dashboard and loading the dashboard results in the following output:
```
Peer URL: https://localhost:4440/iwc/
Applications retrieved via REST Api: Array[2]
Connected to IWC!
Applications retrieved via IWC: Array[0]
Uncaught (in promise) noResource
```

The REST Api is returning the Bread Basket and Air Mail applications as expected, but I'm getting nothing back over the system.api call via the IWC client.

The same seems to be true when I visit https://localhost:4440/iwc/debugger. The "My Apps" tab of the System API is showing 0 applications.

Any ideas?

Thanks!

**Comment by Kevin-K** [comment]
Ah ok I see now what is going on, this is likely a configuration issue for the IWC on the ozp-ansible project. I'm investigating it now, but may need to get some feedback from @arw180 regarding any changes needed on that repo.

As a double check while I attempt to replicate on my end, @theJuddson can you open up the debugger again, under the `General` page there is an `endpoints` tab and see if the table has any data? Also can you open `https://localhost:4440/iwc/js/ozpIwc.conf.js` and post here the contents of it.

As a quick check, I  ran the IWC standalone (with `grunt serve` on the ozp-iwc repo) I get the following (I omitted rows not pertaining to the applications), note this is different than ozp-ansible because its not the same backend. This seems to give me confidence it's not an bug more-so we simply have an invalid config somewhere:

| Resource   | Rel. Link       | Path                               |
|------------|-----------------|------------------------------------|
| Root       |                 | /api                               |
| system.api | ozp:application | http://localhost:13000/api/listing |


@arw180 :
Is the IWC root data still being provided from the `/api` endpoint?

**Comment by Kevin-K** [comment]
So this issue may be needing to be redirected to the ozp-ansible repo, I followed the ozp-ansible build steps but don't have an ozp-iwc in the build. 

If possible try verifying if the `ozpIwc.config.apiRootUrl` is set to `/iwc-api`. in the `js/ozpIwc.conf.js` of your IWC deployment. By default the IWC ships with it set to `/api`, but I believe the ozp-ansible build puts all IWC api data on `/iwc-api` as its root:

Example ozpIwc.conf.js:
```
var ozpIwc = ozpIwc || {};

ozpIwc.config = {
    logLevel: 6,
    backendSupport: true,
    apiRootUrl: "/iwc-api",
    ajaxPoolSize: 3
};
```

**Comment by theJuddson** [comment]
Okay, here's what I'm getting with my VM: 

The endpoints table is mostly empty:
![screen shot 2016-03-01 at 8 28 10 am](https://cloud.githubusercontent.com/assets/10535198/13428364/fb0ff402-df87-11e5-8056-ed20e3550334.png)

And here's my ozpIwc.conf.js:
```javascript
var ozpIwc = ozpIwc || {};

ozpIwc.config = {
    logLevel: 6,
    backendSupport: true,
    apiRootUrl: "/api",
    ajaxPoolSize: 3
};
```

**Comment by theJuddson** [comment]
I'll update that conf file now and post results.

**Comment by theJuddson** [comment]
That has certainly had a positive impact on the System API. The endpoints table now lists the paths as expected:
![screen shot 2016-03-01 at 8 45 00 am](https://cloud.githubusercontent.com/assets/10535198/13428700/237191ec-df8a-11e5-85da-64f0ee786ce4.png)

And the JavaScript I put into the Bread Basket app is now returning results:
```
Peer URL: https://localhost:4440/iwc/
Applications retrieved via REST Api: Array[2]
Connected to IWC!
Applications retrieved via IWC: Array[11]
Success!
```

The only thing that didn't work as expected was the actual launching of the Air Mail listing. Now that my System API is returning data, I'll dig deeper into the launch functionality to see if I can get that working for our custom listings.

**Comment by Kevin-K** [comment]
Cool good to hear, I'll flow that back so the VM config has the proper endpoint, a quick way to check the mismatch on the air mail listing launching would be verifying the reference path being called matches up to the resource name on the system.api table in the debugger.

**Comment by theJuddson** [comment]
It appears as though the launch functionality _is_ working, just not quite as I expected. Maybe you could add some clarity for me. 

I noticed that I was getting a Chrome "Pop-up Blocked" warning, so I enabled all pop-ups for localhost:4440. Now the launch is working, but is opening the listing in a new tab in my browser. Is that the expected functionality? 

I was hoping for it to open as if the user added it to their dashboard manually through the dashboard interface. Is there any way to achieve this through the System API? If not, I may end up writing a JS utility to emulate the user clicking the "add" button, in their dashboard.

**Comment by Kevin-K** [comment]
>Now the launch is working, but is opening the listing in a new tab in my browser. Is that the expected functionality?

Correct, the launcher by default will call window.open which in part opens the application. 

>I was hoping for it to open as if the user added it to their dashboard manually through the dashboard interface. 

Are you referring to the ozp-webtop dashboard? It would be possible to register the webtop to handle the applicaiton opening by adding it to the dashboard, the code isn't in place on the webtop though.

**Short version:** As of current development of ozp-iwc & ozp-webtop, launching with the IWC won't open within the dashboard, the IWC launching logic launches in a new window without any added configuration. Additional JS would be needed to produce custom launching.

---
**More info & how to roll out a custom launching handler in the IWC:**
This functionality isnt a documented feature due to it's primary used internally to the IWC but  just for further info for you, you can register to handle the Intents API intent that the System API calls when launching applications:

```
var launcherRef = new iwc.intents.Reference('/application/vnd.ozp-iwc-launch-data-v1+json/run');
var metaData = {
   label: "My custom launch handler"
};
var onLaunch = function(packet) {

  // url to open
  var url = packet.entity.inFlightIntent.entity.entity.url;

  // 2 manditory query params
  var peer = ozpConfig._busRoot;
  var inFlightIntent = packet.entity.inFlightIntent.resource;

 // do something with the launch data rather than window.open
};
launcherRef.register(metaData,onLaunch);
```

The data that will be returned to onLaunch won't be the prettiest formatted data  but the launching logic will pass both reference to what the bus domain is & an Intents API resource that will give the opening application any launch params passed with the `launch` call.

As long as the `ozpIwc.peer` & `ozpIwc.inFlightIntent` query params are passed onto the resulting url opened (be it new tab, pop-up, iframe, ect) the application should launch & function as if it was opened by the default launcher. This is what the internal launcher calls (util.openWindow essentially is calling window.open and formatting the query params). 
```
util.openWindow(packet.entity.inFlightIntent.entity.entity.url, {
  "ozpIwc.peer": ozpConfig._busRoot,
  "ozpIwc.inFlightIntent": packet.entity.inFlightIntent.resource
});

// This equates to the browser opening a new window at:
// <packet.entity.inFlightIntent.entity.entity.url>?ozpIwc.peer=<ozpConfig._busRoot>&ozpIwc.inFlightIntent=<packet.entity.inFlightIntent.resource>
```

To test you could put the launcherRef snippet within one of your applications, and when you call `launch` on an application reference, a popup window should open prompting for where to route the data for opening. The user can select to have the IWC remember that selection so not to be prompted again during their session.

**Comment by theJuddson** [comment]
Okay, thanks. You've been extremely helpful. I think I'm at a point where I can put something together to achieve the results we want. Assuming I get something reasonable working, I can share the results if this is functionality that could be generally applicable.

I'm closing this ticket, as all of my immediate issues have been resolved.

**Comment by chazzlabs** [comment]
@Kevin-K My team is now looking into IWC and came across this thread.  Our ```ozpIwc.conf.js``` looks like this:

```
var ozpIwc = ozpIwc || {};

ozpIwc.config = {
    logLevel: 6,
    backendSupport: true,
    apiRootUrl: "/api",
    ajaxPoolSize: 3
};

```

In ozp-ansible the ```api_root_url``` variable is still ```/api```.  We can obviously change this in our case, but should this be addressed in ozp-ansible?

**Comment by Kevin-K** [comment]
@chazzlabs correct, there's an open issue on the `ozp-ansible` repo to have that fixed: [#4](https://github.com/aml-development/ozp-ansible/issues/4)
